core/src/main/scala/caliban/execution/Field.scala

```diff
@@ -121,68 +111,72 @@ object Field {
-            fragments
-              .get(name)
-              .foreach { f =>
-                val t = innerType.possibleTypes
-                  .flatMap(_.find(_.name.contains(f.typeCondition.name)))
-                  .orElse(rootType.types.get(f.typeCondition.name))
-                  .getOrElse(fieldType)
-                if (f.selectionSet.nonEmpty) {
-                  loop(f.selectionSet, t).fields.foreach { field =>
-                    val _field =
-                      if (field._condition.isDefined) field
+        case fr @ FragmentSpread(name, directives)                      =>
+          val fields = memoizedFragments.getOrElseUpdate(
+            fr, {
+              val resolvedDirectives = directives.map(resolveDirectiveVariables(variableValues, variableDefinitions))
+
+              if (checkDirectives(resolvedDirectives)) {
+                fragments
+                  .get(name)
+                  .map { f =>
+                    val t = innerType.possibleTypes
+                      .flatMap(_.find(_.name.contains(f.typeCondition.name)))
+                      .orElse(rootType.types.get(f.typeCondition.name))
+                      .getOrElse(fieldType)
+                    loop(f.selectionSet, t).map { field =>
+                      if (field._condition.isDefined) field -> f.typeCondition.name
                       else
                         field.copy(
                           targets = Some(Set(f.typeCondition.name)),
                           _condition = subtypeNames(f.typeCondition.name, rootType)
-                        )
-                    addField(_field, Some(f.typeCondition.name))
+                        )                                   -> f.typeCondition.name
+
+                    }
                   }
-                }
-              }
-          }
-        case InlineFragment(typeCondition, directives, selectionSet)    =>
-          val resolvedDirectives = directives.map(directive =>
-            directive.copy(arguments = resolveVariables(directive.arguments, variableDefinitions, variableValues))
+              } else None
+            }
           )
+          fields.foreach(_.foreach { case (field, name) => addField(field, Some(name)) })
+        case InlineFragment(typeCondition, directives, selectionSet)    =>
+          val resolvedDirectives = directives.map(resolveDirectiveVariables(variableValues, variableDefinitions))
 
           if (checkDirectives(resolvedDirectives)) {
-            val t = innerType.possibleTypes
+            val t      = innerType.possibleTypes
               .flatMap(_.find(_.name.exists(typeCondition.map(_.name).contains)))
               .orElse(typeCondition.flatMap(typeName => rootType.types.get(typeName.name)))
               .getOrElse(fieldType)
-            if (selectionSet.nonEmpty) {
-              val fields = loop(selectionSet, t).fields
-              typeCondition match {
-                case None           => fieldList ++= fields
```

Comment:
It seems to me that this was a bug? If we appended `fields` onto `fieldsList` without incrementing `fieldIndex`, wouldn't that potentially cause the wrong field to be merged?

Reply:
I agree, it looks like a bug!

Reply:
If it's not too difficult, a test case that reproduced the bug would be awesome

Reply:
I'll look into how I can reproduce. It seems that this case would only be reachable with inlined fragments that don't have a type condition. From a quick search, it seems that requires inlined fragments that use the `@skip` or `@include` directives

Reply:
I've managed to create a test case where the wrong fields get combined. Given the query below, with the current code instead of merging `name` + `name`, we end up combining `origin` + `name`. However, I'm not sure how to translate that into an actual issue or get a test to fail. Any thoughts?

```gql
query {
  amos: character(name: "Amos Burton") {
    ... @include(if: true) {
      origin
    }
    name
    name
  }
}
```

Reply:
If you make the resolver of type `Field => ...`, then you can return the list of requested fields, and it would be wrong I think?

Reply:
Okay took me a while but got there eventually 😅 As I mentioned on Discord, the test would fail without the change in the `Field.apply` method

https://github.com/ghostdogpr/caliban/compare/series/2.x...reproduce-inline-fragment-bug

Reply:
Test added in this PR as well